### PR TITLE
Implement alias deprecation in Python

### PR DESCRIPTION
**Pull request by jmborr**
Companion to PR #32525  

**Description of work.**
- [x] New method `const std::string IAlgorithm.aliasDeprecated(const std::string) const;`
- [x] Expose `IAlgorithm.aliasDeprecated` to python so that it can be overriden
- [x] ConfigService option to raise upon use of a deprecated Alias
- [x] implement Alias warnings and optional raise in the algorithm wrapper
- [x] a class decorator `deprecated_alias(date)` will hide implementation of method `aliasDeprecated`
- [x] release notes
- [x] developer documentation on how to rename an algorithm

**To test:**

Run the following script in the workbench:
```python
from mantid.kernel import ConfigService
config = ConfigService.Instance()

config['algorithms.alias.deprecated'] = 'Log'
from mantid.simpleapi import CNCSSuggestTIB, SuggestTibCNCS
x = CNCSSuggestTIB(42)  # use algorithm name
x = SuggestTibCNCS(42)  # use alias
config['algorithms.alias.deprecated'] = 'Raise'
x = SuggestTibCNCS(42)  # use alias, raise RuntimeError


config['algorithms.alias.deprecated'] = 'Log'
from mantid.simpleapi import HYSPECSuggestTIB, SuggestTibHYSPEC
x = HYSPECSuggestTIB(42)  # use algorithm name
x = SuggestTibHYSPEC(42)  # use alias
config['algorithms.alias.deprecated'] = 'Raise'
x = SuggestTibHYSPEC(42)  # use alias, raise RuntimeError
```

There is no associated issue (see [PL100](https://code.ornl.gov/sns-hfir-scse/planning/-/issues/100))

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Properties/Mantid.properties.template [hunk 1]
```diff
@@ -186,6 +186,12 @@ loading.multifilelimit =
 # Hide algorithms that use a Property Manager by default.
 algorithms.categories.hidden=Workflow\\Inelastic\\UsesPropertyManager;Workflow\\SANS\\UsesPropertyManager;DataHandling\\LiveData\\Support;Deprecated;Utility\\Development;Remote
 
+# Response upon invoking an algorithm with its alias, when the alias is deprecated.
+# Allowed values are:
+#   "Warn": log a warning indicating the alias is deprecated, along with the name to be used
+#   "Raise": raise a RuntimeError if the deprecated deadline has been met
+algorithms.alias.deprecated = Warn
```
Comment (peterfpeterson) [inline Framework/Properties/Mantid.properties.template:193]: Please be more linux-y and change these to be all lowercase (i.e. `warn` and `raise`)

File: Framework/PythonInterface/mantid/simpleapi.py [hunk 2]
```diff
@@ -983,6 +1008,14 @@ def __call__(self, *args, **kwargs):
                 If both startProgress and endProgress are supplied they will
                 be used.
                 """
+
+                # Check at runtime whether to throw upon alias deprecation.
+                if self._alias and self._alias.deprecated:
+                    deprecated = parse_date(self._alias.deprecated) < datetime.datetime.today()
+                    deprecated_action = ConfigService.Instance().get('algorithms.alias.deprecated', 'Warn').lower()
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/simpleapi.py:1015]: Even you want it all lowercase!

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Other than changing the case of the property value and missing release notes, I think this is ready.

File: Framework/Properties/Mantid.properties.template (hunk 1 above)
Comment (jmborr) [reply Framework/Properties/Mantid.properties.template:193]: I'm following the convention of [PropertiesFile](https://github.com/mantidproject/mantid/blob/master/docs/source/concepts/PropertiesFile.rst). Looks like most options have their initials capitalized

File: Framework/PythonInterface/mantid/simpleapi.py (hunk 2 above)
Comment (jmborr) [reply Framework/PythonInterface/mantid/simpleapi.py:1015]: I'm allowing for a user forgetting to capitalize the initial 🙂

File: Framework/Properties/Mantid.properties.template (hunk 1 above)
Comment (peterfpeterson) [reply Framework/Properties/Mantid.properties.template:193]: Reminding us that the first mantid developers were on windows :(

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**